zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/IPAuthenticationProvider.java

```diff
@@ -128,4 +131,18 @@ public boolean isValid(String id) {
         return true;
     }
 
+    /**
+     * Returns the HTTP(s) client IP address
+     * @param request HttpServletRequest
+     * @return IP address
+     */
+    public static String getClientIPAddress(final HttpServletRequest request) {
+        // to handle the case that a HTTP(s) client connects via a proxy or load balancer
+        final String xForwardedForHeader = request.getHeader(X_FORWARDED_FOR_HEADER_NAME);
+        if (xForwardedForHeader == null) {
+            return request.getRemoteAddr();
+        }
+        // the format of the field is: X-Forwarded-For: client, proxy1, proxy2 ...
+        return new StringTokenizer(xForwardedForHeader, ",").nextToken().trim();
+    }
```

Comment:
I don't think this method belongs to this class. `IPAuthenticationProvider` is responsible for authentication ZooKeeper client by the client IP address. You don't use any member of the class, neither instantiate it (static method), so please move it where the usage is (StreamOutputter.java).

Reply:
This method is in `IPAuthenticationProvider `as it will be used by `IPAuthenticationProvider` when providing IP auth support for admin server APIs. 

`IPAuthenticationProvider` will be enhanced for authenticating both Zookeeper client and Admin Server APIs by the client IP address.

Reply:
Fine. Next time please add a refactoring to the upcoming patch instead. Looking at this patch alone means that you break the encapsulation principle.

Reply:
Yep